pallets/manta-sbt/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+            // Checks that it is indeed a to_private post with a value of 1
+            ensure!(
+                post.sources.len() == 1
+                    && post.sender_posts.is_empty()
+                    && post.receiver_posts.len() == 1
+                    && post.sinks.is_empty(),
+                Error::<T>::NoSenderLedger
+            );
+            // Checks that value is one, this is defensive as value is not used for SBT.
+            ensure!(
+                post.sources.first() == Some(&ENCODED_ONE),
+                Error::<T>::ValueNotOne
+            );
+
+            let (start_id, end_id) =
+                ReservedIds::<T>::get(&origin).ok_or(Error::<T>::NotReserved)?;
+            let asset_id: StandardAssetId = post
+                .asset_id
+                .and_then(id_from_field)
+                .ok_or(Error::<T>::InvalidAssetId)?;
+            // Ensure asset id is correct, only a single unique asset_id mapped to account is valid
+            ensure!(asset_id == start_id, Error::<T>::InvalidAssetId);
+
+            // Updates SbtMetadata
+            T::UpdateMetadata::update_metadata(&asset_id, AssetMetadata::SBT(metadata))?;
+            let increment_start_id = start_id
+                .checked_add(One::one())
+                .ok_or(ArithmeticError::Overflow)?;
+
+            // If `ReservedIds` are all used remove from storage, otherwise increment the next `AssetId` to be used next time for minting SBT
+            if increment_start_id > end_id {
+                ReservedIds::<T>::remove(&origin)
+            } else {
+                ReservedIds::<T>::insert(&origin, (increment_start_id, end_id))
+            }
+
+            T::Ledger::post_transaction(None, vec![origin.clone()], vec![], *post, AssetType::SBT)?;
+            Self::deposit_event(Event::<T>::MintSbt {
+                source: origin,
+                asset: asset_id,
+            });
+            Ok(().into())
+        }
+
+        /// Reserves AssetIds to be used subsequently in `to_private` above.
+        ///
+        /// Increments AssetManager's AssetId counter.
+        #[pallet::call_index(1)]
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::reserve_sbt())]
+        #[transactional]
+        pub fn reserve_sbt(origin: OriginFor<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // Charges fee to reserve AssetIds
+            <T as pallet::Config>::Currency::transfer(
+                &who,
+                &Self::account_id(),
+                T::ReservePrice::get(),
+                ExistenceRequirement::KeepAlive,
+            )?;
+
+            // Reserves uniques AssetIds to be used later to mint SBTs
+            let asset_id_range: Vec<StandardAssetId> = (0..T::MintsPerReserve::get())
+                .map(|_| T::UpdateMetadata::create_asset(AssetMetadata::SBT(BoundedVec::default())))
+                .collect::<Result<Vec<StandardAssetId>, _>>()?;
```

Comment:
Maybe `collect` can be remove here.
Memory will be allocated once `collect` is called, but iterator won't due to laziness. And seems you just iterate `asset_id_range`.

Reply:
I think collect is needed as it changes the Range value to the incremented `NextSbtId` in the `map`.

Could be mistaken but I don't see how to, could you show me snippit if I am missing something

Reply:
`MintsPerReserve` is really small, so it's fine that we still use `collect`.